### PR TITLE
Federation Data Exchange for FE2

### PR DESCRIPTION
**Pull request by quadcopterman**
Here is the implementation of the data exchange protocol for FE2 Android. The data exchange protocol is defined in issue #1802

### CODE REVIEW

File: fe2-android/app/src/main/java/com/github/dedis/popstellar/model/network/method/message/data/federation/FederationResult.kt [hunk 1]
```diff
@@ -0,0 +1,77 @@
...
+    val challenge: MessageGeneral,
+) : Data {
+
+  init {
+    when (status) {
+      "failure" -> {
+        require(reason != null) { "Reason must be provided for failure status." }
+        require(publicKey == null) { "Public key must be null for failure status." }
+      }
+      "success" -> {
```
Comment (matteosz) [inline fe2-android/app/src/main/java/com/github/dedis/popstellar/model/network/method/message/data/federation/FederationResult.kt:32]: Perhaps it would be better to define these ("failure" and "success") as static constants, so in case we want to change them in future it's just 1 place to modify

File: fe2-android/app/src/main/java/com/github/dedis/popstellar/repository/LinkedOrganizationsRepository.kt [hunk 2]
```diff
@@ -35,6 +34,36 @@ class LinkedOrganizationsRepository @Inject constructor() {
...
+  }
+
+  fun updateAndNotifyLinkedLao(
+      laoId: String,
+      otherLaoId: String,
+      tokens: Array<String>,
+      rollCallId: String
+  ) {
+    addLinkedLao(laoId, otherLaoId, tokens)
+    newTokensNotifyFunction?.invoke(laoId, otherLaoId, rollCallId, tokens)
```
Comment (matteosz) [inline fe2-android/app/src/main/java/com/github/dedis/popstellar/repository/LinkedOrganizationsRepository.kt:50]: If you take a look at other repos (any one, e.g. rollcallrepo) you'd see that in general this is not the preferred way to notify modifications on a given data structure, but that we use java rx constructs (subject, observables and so on)

File: fe2-android/app/src/main/java/com/github/dedis/popstellar/ui/lao/federation/LinkedOrganizationsFragment.kt [hunk 3]
```diff
@@ -59,6 +59,19 @@ class LinkedOrganizationsFragment : Fragment() {
...
+    linkedOrganizationsViewModel.doWhenLinkedLaosIsUpdated { laoId, laoMap ->
+      if (laoId == laoViewModel.laoId) {
+        CoroutineScope(Dispatchers.Main).launch {
+          val currentLaos = laoMap.keys
+          displayLinkedOrganizations(currentLaos)
+        }
+      }
+    }
+
+    linkedOrganizationsViewModel.setLinkedLaosNotifyFunction()
```
Comment (matteosz) [inline fe2-android/app/src/main/java/com/github/dedis/popstellar/ui/lao/federation/LinkedOrganizationsFragment.kt:74]: Why calling display first with the current list and then putting a custom and complex observer that uses coroutines and not directly using a simple observer? This relates to my previous comment on how to deal with modifications using observables and subjects

File: fe2-android/app/src/main/java/com/github/dedis/popstellar/ui/lao/socialmedia/SocialMediaViewModel.kt [hunk 4]
```diff
@@ -225,36 +227,50 @@ constructor(
...
+                .sorted(
+                    Comparator.comparing { chirp: Chirp? ->
+                      if (chirp != null) -chirp.timestamp else 0
+                    })
+                .collect(Collectors.toList())
+          }
+      // We want to observe these changes on the main thread such that any modification done
+      // to the view are done on the thread. Otherwise, the app might crash
+      return combinedObservables.observeOn(schedulerProvider.mainThread())
+    }
```
Comment (matteosz) [inline fe2-android/app/src/main/java/com/github/dedis/popstellar/ui/lao/socialmedia/SocialMediaViewModel.kt:270]: This getter will be executed every time we access that field, which looks very heavy as getter method. You could precompute this variable in initialization, such that this function is only called once

File: fe2-android/app/src/main/java/com/github/dedis/popstellar/utility/handler/data/LinkedOrganizationsHandler.kt [hunk 5]
```diff
@@ -24,4 +31,84 @@ constructor(
...
+   */
+  @Throws(UnknownLaoException::class)
+  fun handleResult(context: HandlerContext, result: FederationResult) {
+    if (result.isSuccess()) {
+      if (result.challenge.data == linkedOrgRepo.getChallenge() &&
+          linkedOrgRepo.otherLaoId != null) {
+        val laoId = context.channel.extractLaoId()
+        linkedOrgRepo.addLinkedLao(laoId, linkedOrgRepo.otherLaoId!!, arrayOf())
+        laoRepo.addDisposable(
+            context.messageSender
```
Comment (matteosz) [inline fe2-android/app/src/main/java/com/github/dedis/popstellar/utility/handler/data/LinkedOrganizationsHandler.kt:49]: Maybe for better separability it would be nice to have an ad-hoc disposable in the linked org repo, without using the lao one

File: fe2-android/app/src/test/ui/robolectric/com/github/dedis/popstellar/ui/lao/federation/LinkedOrganizationsFragmentOrganizerTest.kt
Comment (matteosz) [inline fe2-android/app/src/test/ui/robolectric/com/github/dedis/popstellar/ui/lao/federation/LinkedOrganizationsFragmentOrganizerTest.kt:1]: Could you add a test when also there are some organizations (so not just the empty case)?

**Review by matteosz — CHANGES_REQUESTED** [review]
Great work!

File: fe2-android/app/src/main/java/com/github/dedis/popstellar/model/network/method/message/data/federation/FederationResult.kt (hunk 1 above)
Comment (quadcopterman) [reply fe2-android/app/src/main/java/com/github/dedis/popstellar/model/network/method/message/data/federation/FederationResult.kt:32]: Ok, I'll change that

File: fe2-android/app/src/main/java/com/github/dedis/popstellar/repository/LinkedOrganizationsRepository.kt (hunk 2 above)
Comment (quadcopterman) [reply fe2-android/app/src/main/java/com/github/dedis/popstellar/repository/LinkedOrganizationsRepository.kt:50]: Well the thing is that I tried using observables and for some reason I had many problems with the UI not updating or weird other issues... That's why in the end I changed to using callbacks, I know it's a bit more complex but at least it's working

File: fe2-android/app/src/main/java/com/github/dedis/popstellar/utility/handler/data/LinkedOrganizationsHandler.kt (hunk 5 above)
Comment (quadcopterman) [reply fe2-android/app/src/main/java/com/github/dedis/popstellar/utility/handler/data/LinkedOrganizationsHandler.kt:49]: Yes, I'll do that

File: fe2-android/app/src/main/java/com/github/dedis/popstellar/repository/LinkedOrganizationsRepository.kt (hunk 2 above)
Comment (matteosz) [reply fe2-android/app/src/main/java/com/github/dedis/popstellar/repository/LinkedOrganizationsRepository.kt:50]: Sure, we can keep it as is then

File: fe2-android/app/src/main/java/com/github/dedis/popstellar/ui/lao/socialmedia/SocialMediaViewModel.kt (hunk 4 above)
Comment (matteosz) [reply fe2-android/app/src/main/java/com/github/dedis/popstellar/ui/lao/socialmedia/SocialMediaViewModel.kt:270]: Yeah sorry I was missing the context of execution of that function, now I looked it up better and it's how it's supposed to work, even if it's not very efficient and perhaps in the future we could change it

**Review by Kaz-ookid — CHANGES_REQUESTED** [review]
Matteo's comments cover all the important parts but I have one last request:
FederationResult and TokensExchange are new message data, that require validation of inputs.

You should add the validation using MessageValidation.validate() to avoid having technical debt, as total validation is a goal documented in #1751. You should also update this Issue and add the new data in the expandable columns :)

You can check in #1952 how to use validation, and how to create simple tests to check the constructor.

File: fe2-android/app/src/main/java/com/github/dedis/popstellar/model/network/method/message/data/federation/FederationResult.kt [hunk 6]
```diff
@@ -0,0 +1,84 @@
...
+ */
+(
+    val status: String,
+    val reason: String? = null,
+    @SerializedName("public_key") val publicKey: String? = null,
+    val challenge: MessageGeneral,
+) : Data {
+
+  init {
+    when (status) {
```
Comment (Kaz-ookid) [inline fe2-android/app/src/main/java/com/github/dedis/popstellar/model/network/method/message/data/federation/FederationResult.kt:28]: This looks like input validation, maybe you should complete MessageValidator class to check a federation status? Validation would then just be a clear oneliner 👍

File: fe2-android/app/src/main/java/com/github/dedis/popstellar/model/network/method/message/data/federation/TokensExchange.kt [hunk 7]
```diff
@@ -0,0 +1,59 @@
...
+ */
+(
+    @SerializedName("lao_id") val laoId: String,
+    @SerializedName("roll_call_id") val rollCallId: String,
+    val tokens: Array<String>,
+    val timestamp: Long
+) : Data {
+
+  init {
+    MessageValidator.verify().isBase64(rollCallId, "rollCallId")
```
Comment (Kaz-ookid) [inline fe2-android/app/src/main/java/com/github/dedis/popstellar/model/network/method/message/data/federation/TokensExchange.kt:27]: isBase64 (as every other functions in MessageValidator) returns the builder so you can write : 

`MessageValidator.verify().isBase64(rollCallId, "rollCallId").validPastTimes(timestamp)`

File: fe2-android/app/src/main/java/com/github/dedis/popstellar/utility/handler/data/LinkedOrganizationsHandler.kt [hunk 8]
```diff
@@ -76,15 +77,15 @@ constructor(
     // LAO. This might be changed in the future (making a pop-up asking the user if he/she wants
     // to subscribe to that)
     tokenExchange.tokens.forEach { t ->
-      laoRepo.addDisposable(
+      disposables.add(
```
Comment (matteosz) [inline fe2-android/app/src/main/java/com/github/dedis/popstellar/utility/handler/data/LinkedOrganizationsHandler.kt]: Using the disposable from the handler doesn't give you the opportunity to clear the added disposables when closing a lao or the application. Take a look at the usage of disposables in the RollCallRepository or any other repository.

File: fe2-android/app/src/main/java/com/github/dedis/popstellar/model/network/method/message/data/federation/TokensExchange.kt (hunk 7 above)
Comment (Kaz-ookid) [reply fe2-android/app/src/main/java/com/github/dedis/popstellar/model/network/method/message/data/federation/TokensExchange.kt:27]: given the documentation of the protocol :

<details>
<summary><code>consensus</code></summary>

```
// ../protocol/query/method/message/data/dataFederationTokensExchange.json
{
  "$schema": "http://json-schema.org/draft-07/schema#",
  "$id": "https://raw.githubusercontent.com/dedis/popstellar/master/protocol/query/method/message/data/dataFederationTokensExchange.json",
  "description": "Sent by an organizer client to its server, to broadcast the Pop tokens",
  "type": "object",
  "properties": {
    "object": {
      "const": "federation"
    },
    "action": {
      "const": "tokens_exchange"
    },
    "lao_id": {
      "type": "string",
      "contentEncoding": "base64",
      "$comment": "Hash : HashLen(organizer, creation, name)"
    },
    "roll_call_id": {
      "type": "string",
      "contentEncoding": "base64",
      "$comment": "last roll call id"
    },
    "tokens": {
      "description": "[Array[Base64String]] list of Pop tokens",
      "type": "array",
      "uniqueItems": true,
      "items": {
        "type": "string",
        "contentEncoding": "base64"
      },
      "$comment": "List must be sorted according to byte encoding: -,0...9,A...Z,_,a...z"
    },
    "timestamp": {
      "type": "integer",
      "description": "[Timestamp] of the tokens' exchange",
      "minimum": 0
    }
  },
  "additionalProperties": false,
  "required": [
    "object",
    "action",
    "lao_id",
    "roll_call_id",
    "tokens",
    "timestamp"
  ]
}
```


</details>

- It looks like that laoId should be checked for `notEmptybase64`.
- I am not sure if an empty rollCallId makes sense, if not then switch to `notEmptybase64 `as well.
- Tokens look like they individually need to be checked as well (they should be base64, maybe we wan't them not empty as well?). Also they should be checked for uniqueness. Is a null or empty array acceptable? This maybe needs to be defined.

File: fe2-android/app/src/test/unit/java/com/github/dedis/popstellar/model/network/method/message/data/federation/FederationResultTest.kt [hunk 9]
```diff
@@ -0,0 +1,137 @@
...
+        }
+        assert(exception.message == "Public key must be provided for success status.")
+    }
+
+    @Test
+    fun invalidSuccessTest2() {
+        val exception = assertThrows(IllegalArgumentException::class.java) {
+            FederationResult(SUCCESS, publicKey = PK.encoded, reason = "reason", challenge = MG_CHALLENGE)
+        }
+        assert(exception.message == "Reason must be null for success status.")
```
Comment (Kaz-ookid) [inline fe2-android/app/src/test/unit/java/com/github/dedis/popstellar/model/network/method/message/data/federation/FederationResultTest.kt:105]: maybe we want to add those strings in Constants to avoid failing if the message comes to change

File: fe2-android/app/src/main/java/com/github/dedis/popstellar/utility/handler/data/LinkedOrganizationsHandler.kt (hunk 8 above)
Comment (quadcopterman) [reply fe2-android/app/src/main/java/com/github/dedis/popstellar/utility/handler/data/LinkedOrganizationsHandler.kt]: Oh ok, then should I create the disposables in LinkedOrganizationsRepository (as done in RollCallRepository) and then call it from the LinkedOrganizationsHandler ? Or should I do it directly in the handler ?

File: fe2-android/app/src/test/unit/java/com/github/dedis/popstellar/model/network/method/message/data/federation/TokensExchangeTest.kt [hunk 10]
```diff
@@ -0,0 +1,77 @@
...
+
+import com.github.dedis.popstellar.model.network.method.message.data.Action
+import com.github.dedis.popstellar.model.network.method.message.data.Objects
+import com.github.dedis.popstellar.model.objects.Lao
+import com.github.dedis.popstellar.testutils.Base64DataUtils
+import org.junit.Assert
+import org.junit.Test
+import java.time.Instant
+
+class TokensExchangeTest {
```
Comment (Kaz-ookid) [inline fe2-android/app/src/test/unit/java/com/github/dedis/popstellar/model/network/method/message/data/federation/TokensExchangeTest.kt:11]: don't forget to add constructors succeed and failing tests as in the other class

**Review by Kaz-ookid — CHANGES_REQUESTED** [review]
Great! Just a few last steps in message validation 👍

File: fe2-android/app/src/main/java/com/github/dedis/popstellar/utility/handler/data/LinkedOrganizationsHandler.kt (hunk 8 above)
Comment (matteosz) [reply fe2-android/app/src/main/java/com/github/dedis/popstellar/utility/handler/data/LinkedOrganizationsHandler.kt]: Yep you can do it from the repository as for roll calls and keep it consistent

**Review by matteosz — APPROVED** [review]
